planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -1,217 +1,116 @@
-      System.out.println("ERROR in concat().");
-    }
-    return GeoUtils.JTS_FACTORY.createLineString(coordinates.toArray(new Coordinate[0]));
-  }
-
   private void merge() {
-    for (var point : graph.keySet()) {
-      var node = graph.get(point);
+    for (var node : output) {
       if (node.getEdges().size() == 2) {
-
-        var A = node.getEdges().get(0);
-        var B = node.getEdges().get(1);
-
-        graph.get(A.getStartPoint()).removeEdge(A);
-        graph.get(A.getEndPoint()).removeEdge(A);
-
-        graph.get(B.getStartPoint()).removeEdge(B);
-        graph.get(B.getEndPoint()).removeEdge(B);
-
-        var C = concat(A, B);
-
-        graph.get(C.getStartPoint()).addEdge((LineString) C.copy());
-        graph.get(C.getEndPoint()).addEdge((LineString) C.copy());
+        Edge a = node.getEdges().getFirst();
+        Edge b = node.getEdges().get(1);
+        node.getEdges().clear();
+        List<Coordinate> coordinates = new ArrayList<>();
+        coordinates.addAll(a.coordinates.reversed());
+        coordinates.addAll(b.coordinates.subList(1, b.coordinates.size()));
+        Edge c = new Edge(a.to, b.to, coordinates, a.length + b.length);
+        a.to.removeEdge(a.reversed);
+        b.to.removeEdge(b.reversed);
+        a.to.addEdge(c);
+        b.to.addEdge(c.reversed);
       }
     }
-
-    // remove nodes that do not have edges attached anymore
-    graph.entrySet().removeIf(entry -> entry.getValue().getEdges().size() == 0);
   }
 
-  private void removeLoops(double loopMinLength) {
-
-    var points = new ArrayList<>(graph.keySet());
-    for (var currentPoint : points) {
-
-      var node = graph.get(currentPoint);
-      if (node == null) {
-        continue;
-      }
-
-      var edges = new ArrayList<>(node.getEdges());
-      while (edges.size() > 0) {
-        var currentEdge = edges.get(0);
-        edges.remove(0);
-
-        if (!node.getEdges().contains(currentEdge)) {
-          continue;
-        }
-
-        var A = currentEdge.getStartPoint();
-        var B = currentEdge.getEndPoint();
-
-        var end = currentPoint.equals(A) ? B : A;
-        var allPaths = findAllPaths(currentPoint, end, loopMinLength);
-
+  private void removeLoops() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        var allPaths = findAllPaths(edge.from, edge.to, loopMinLength);
         if (allPaths.size() > 1) {
           for (var path : allPaths.subList(1, allPaths.size())) {
-            if (path.size() > 0) {
-              var firstEdge = path.get(0);
-              var startNode = graph.get(firstEdge.getStartPoint());
-              if (startNode != null) {
-                startNode.removeEdge(firstEdge);
-              }
-              var endNode = graph.get(firstEdge.getEndPoint());
-              if (endNode != null) {
-                endNode.removeEdge(firstEdge);
-              }
+            for (var toRemove : path) {
+              toRemove.remove();
```

Comment:
Here the old version just remove the first edge of a loop path from the graph, and not all edges of the path. So that was sort of like cutting the loop open, later line dropping by length would remove dangling lines.